modules/auxiliary/cloud/aws/enum_ssm.rb

```diff
@@ -0,0 +1,156 @@
+        OptString.new('ACCESS_KEY_ID', [true, 'AWS Access Key ID (eg. "AKIAXXXXXXXXXXXXXXXX")', '']),
+        OptString.new('SECRET_ACCESS_KEY', [true, 'AWS Secret Access Key (eg. "CA1+XXXXXXXXXXXXXXXXXXXXXX6aYDHHCBuLuV79")', ''])
+      ]
+    )
+  end
+
+  def handle_aws_errors(error)
+    if error.class.module_parents.include?(Aws)
+      fail_with(Failure::UnexpectedReply, error.message)
+    else
+      raise error
+    end
+  end
+
+  def run
+    credentials = ::Aws::Credentials.new(datastore['ACCESS_KEY_ID'], datastore['SECRET_ACCESS_KEY'])
+    vprint_status "Checking #{datastore['REGION']}..."
+    client = ::Aws::SSM::Client.new(
+      region: datastore['REGION'],
+      credentials: credentials
+    )
+    inv_params = {
+      filters: [
+        {
+          key: 'AWS:InstanceInformation.InstanceStatus',
+          values: ['Terminated'],
+          type: 'NotEqual'
+        },
+        {
+          key: 'AWS:InstanceInformation.ResourceType',
+          values: ['EC2Instance'],
+          type: 'Equal'
+        }
+      ]
+    }
+
+    if datastore['FILTER_EC2_ID']
+      inv_params[:filters] << {
+        key: 'AWS:InstanceInformation.InstanceId',
+        values: [datastore['FILTER_EC2_ID']],
+        type: 'Equal'
+      }
+    end
+
+    ssm_ec2 = client.get_inventory(inv_params).entities.map { |e| e.data['AWS:InstanceInformation'].content }.flatten
+    ssm_ec2 = ssm_ec2[0...datastore['LIMIT']] if datastore['LIMIT']
+    ssm_ec2.each do |ssm_host|
+      report_host(
+        host: ssm_host['IpAddress'],
+        os_flavor: ssm_host['PlatformName'],
+        os_name: ssm_host['PlatformType'],
+        os_sp: ssm_host['PlatformVersion'],
+        name: ssm_host['ComputerName'],
+        comments: "ec2-id: #{ssm_host['InstanceId']}"
+      )
+      report_note(
+        host: ssm_host['IpAddress'],
+        type: ssm_host['AgentType'],
+        data: ssm_host['AgentVersion']
+      )
+      vprint_good("Found AWS SSM host #{ssm_host['InstanceId']} (#{ssm_host['ComputerName']}) - #{ssm_host['IpAddress']}")
+      next unless datastore['CreateSession']
+
+      socket = get_ssm_socket(client, ssm_host['InstanceId'])
+      sess = Msf::Sessions::AwsSsmCommandShellBind.new(socket.lsock, { datastore: datastore, aws_ssm_host_info: ssm_host })
+
+      start_session(self, sess.info, datastore, false, socket.lsock, sess)
+    end
+  rescue Seahorse::Client::NetworkingError => e
+    print_error e.message
+    print_error "Confirm access to #{datastore['REGION']} with provided credentials"
+  rescue StandardError => e
+    handle_aws_errors(e)
+  end
+
+  def get_ssm_socket(client, ec2_id)
+    # Verify the connection params and availability of instance
+    inv_params = {
+      filters: [
+        {
+          key: 'AWS:InstanceInformation.InstanceId',
+          values: [ec2_id],
+          type: 'Equal'
+        }
+      ]
+    }
+    inventory = client.get_inventory(inv_params)
```

Comment:
Not a blocker; it feels like this n+1 query could be removed. i.e. the additional lookup of `get_inventory` for every connection opened feels like it could be avoided with some code shuffling in the parent loop. I haven't verified this claim though.